build_stage_repository

```diff
@@ -172,40 +166,143 @@ def packages_from_comps(comps):
+        else:
+            print(f"Skipping {repository}/{name}. Already downloaded.")
+
+
+def handle_args():
+    parser = argparse.ArgumentParser(description='Generate a stage repository')
+    parser.add_argument(
+        'collection',
+        help='Repository to generate for (e.g. foreman)'
+    )
+    parser.add_argument(
+        'version',
+        help='Version to generate the repository for (e.g. nightly)'
+    )
+    parser.add_argument(
+        'dist',
+        help='Dist to generate repository for (e.g. el8)'
+    )
+
+    return parser.parse_args()
+
+
 def main():
-    try:
-        collection = sys.argv[1]
-        version = sys.argv[2]
-        dist = sys.argv[3]
-        arch = 'x86_64'
-    except IndexError:
-        raise SystemExit(f"Usage: {sys.argv[0]} collection version os")
+    args = handle_args()
 
-    base_dir = 'tmp'
-    rpm_sync_dir = f"{base_dir}/rpms"
-    srpm_sync_dir = f"{base_dir}/srpms"
+    collection = args.collection
+    version = args.version
+    dist = args.dist
+    arch = 'x86_64'
 
-    stage_dir = f"{base_dir}/{collection}/{version}/{dist}/"
+    base_dir = 'tmp'
+    stage_dir = f"{base_dir}/{collection}/{version}/{dist}"
     rpm_dir = f"{stage_dir}/{arch}"
     srpm_dir = f"{stage_dir}/source"
 
-    if not os.path.exists(rpm_sync_dir):
-        os.makedirs(rpm_sync_dir)
-
-    if not os.path.exists(srpm_sync_dir):
-        os.makedirs(srpm_sync_dir)
-
     if not os.path.exists(rpm_dir):
         os.makedirs(rpm_dir)
 
     if not os.path.exists(srpm_dir):
         os.makedirs(srpm_dir)
 
-    sync_copr_repository(collection, version, rpm_sync_dir, dist, arch)
-    sync_copr_repository(collection, version, srpm_sync_dir, dist, arch, source=True)
+    if version != 'nightly':
+        sync_prod_repository(collection, version, rpm_dir, dist, arch)
+        sync_prod_repository(collection, version, srpm_dir, dist, 'source')
+    else:
+        create_repository(rpm_dir)
+        create_repository(srpm_dir)
 
-    move_rpms_from_copr_to_stage(collection, version, srpm_sync_dir, srpm_dir, dist, arch, source=True)
-    move_rpms_from_copr_to_stage(collection, version, rpm_sync_dir, rpm_dir, dist, arch)
+    packages_to_include = packages_from_comps(comps(collection, version, dist))
+    copr_repo = copr_repository(collection, version, dist, arch)
+    copr_package_urls = copr_repository_urls(copr_repo)
+
+    rpm_diff = compare_repositories(copr_repo, rpm_dir)
+    packages = parse_repodiff(rpm_diff)
+    download_copr_packages(packages, copr_package_urls, copr_repo, rpm_dir, packages_to_include)
+    filter_packages(rpm_dir, packages_to_include)
+
+    srpm_diff = compare_repositories(copr_repo, srpm_dir, True)
+    srpm_packages = parse_repodiff(srpm_diff)
+    download_copr_packages(srpm_packages, copr_package_urls, copr_repo, srpm_dir, packages_to_include)
+    filter_packages(srpm_dir, packages_to_include)
 
     create_repository(rpm_dir)
```

Comment:
this will erase "old" module metadata from `rpm_dir` (if there is any, e.g. when `sync_prod_repository` fetched some).

in https://github.com/theforeman/foreman-infra/blob/376d691f2efca2d0bae8eb594c10cfd51e2117f3/puppet/modules/web/files/deploy-yumrepo.sh#L51-L64 there is an example of how we (today) merge old module metadata, and I think this should be somehow transferred here, as otherwise we'll end up with a repo that has RPMs that are not referenced by any version of a module

Reply:
I am not seeing how we would end up with RPMs not in the module metadata. We download production, calculate the diff, download the diff, create a repo and then add module metadata based on all packages in the repository.

Reply:
Again me mixing up flows. The result feeds in all rpms, not only the new ones.

(What does it tell about me, or the code, that I misread it in two places?)